### ISSUE TITLE
When sqs queue is empty, EKS pods throwsEXCEPTION: java.lang.NoClassDefFoundError: io/netty/util/internal/CleanerJava9$2

### ISSUE THREAD

**Issue by sashjena** [aws/aws-sdk-java#2875]
### Describe the bug

[2022-11-08T10:01:33.820+0000]   [       ]  [DEBUG] [id: 0xef2e81e5, L:/10.38.86.232:48790 - R:sqs.us-west-2.amazonaws.com/10.38.95.58:443] FLUSH
[2022-11-08T10:01:33.820+0000]   [       ]  [DEBUG] [id: 0xb9ea13b3, L:/10.38.86.232:54632 - R:sqs.us-west-2.amazonaws.com/10.38.15.117:443] FLUSH
[2022-11-08T10:01:33.820+0000]   [       ]  [DEBUG] [id: 0x5dd766f1, L:/10.38.86.232:48792 - R:sqs.us-west-2.amazonaws.com/10.38.95.58:443] FLUSH
[2022-11-08T10:01:33.820+0000]   [       ]  [DEBUG] [id: 0xef2e81e5, L:/10.38.86.232:48790 - R:sqs.us-west-2.amazonaws.com/10.38.95.58:443] WRITE, EmptyLastHttpContent, 0B
[2022-11-08T10:01:33.820+0000]   [       ]  [DEBUG] [id: 0x5dd766f1, L:/10.38.86.232:48792 - R:sqs.us-west-2.amazonaws.com/10.38.95.58:443] WRITE, EmptyLastHttpContent, 0B
[2022-11-08T10:01:33.820+0000]   [       ]  [DEBUG] [id: 0xb9ea13b3, L:/10.38.86.232:54632 - R:sqs.us-west-2.amazonaws.com/10.38.15.117:443] WRITE, EmptyLastHttpContent, 0B
[2022-11-08T10:01:33.820+0000]   [       ]  [DEBUG] [id: 0xb9ea13b3, L:/10.38.86.232:54632 - R:sqs.us-west-2.amazonaws.com/10.38.15.117:443] FLUSH
[2022-11-08T10:01:33.820+0000]   [       ]  [DEBUG] [id: 0xef2e81e5, L:/10.38.86.232:48790 - R:sqs.us-west-2.amazonaws.com/10.38.95.58:443] FLUSH
[2022-11-08T10:01:33.820+0000]   [       ]  [DEBUG] [id: 0x5dd766f1, L:/10.38.86.232:48792 - R:sqs.us-west-2.amazonaws.com/10.38.95.58:443] FLUSH
[2022-11-08T10:01:38.826+0000]   [       ]  [DEBUG] [id: 0xb9ea13b3, L:/10.38.86.232:54632 - R:sqs.us-west-2.amazonaws.com/10.38.15.117:443] EXCEPTION: java.lang.NoClassDefFoundError: io/netty/util/internal/CleanerJava9$2
java.lang.NoClassDefFoundError: io/netty/util/internal/CleanerJava9$2
	at io.netty.util.internal.CleanerJava9.freeDirectBufferPrivileged(CleanerJava9.java:98) ~[netty-common-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.util.internal.CleanerJava9.freeDirectBuffer(CleanerJava9.java:93) ~[netty-common-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.util.internal.PlatformDependent.freeDirectBuffer(PlatformDependent.java:523) ~[netty-common-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.buffer.UnpooledDirectByteBuf.freeDirect(UnpooledDirectByteBuf.java:111) ~[netty-buffer-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeDirectByteBuf.freeDirect(UnpooledByteBufAllocator.java:223) ~[netty-buffer-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.buffer.UnpooledDirectByteBuf.deallocate(UnpooledDirectByteBuf.java:646) ~[netty-buffer-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.buffer.AbstractReferenceCountedByteBuf.handleRelease(AbstractReferenceCountedByteBuf.java:111) ~[netty-buffer-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.buffer.AbstractReferenceCountedByteBuf.release(AbstractReferenceCountedByteBuf.java:101) ~[netty-buffer-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:288) ~[netty-codec-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251) ~[netty-transport-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[netty-transport-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[netty-transport-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[netty-transport-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1373) ~[netty-handler-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1236) ~[netty-handler-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1285) ~[netty-handler-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:510) ~[netty-codec-4.1.79.Final.jar!/:4.1.79.

### Expected Behavior

SQS queue must wait till the subsequent message appears.

### Current Behavior

Throwing below erroe when queue is empty

[2022-11-08T10:01:38.826+0000]   [       ]  [DEBUG] [id: 0xb9ea13b3, L:/10.38.86.232:54632 - R:sqs.us-west-2.amazonaws.com/10.38.15.117:443] EXCEPTION: java.lang.NoClassDefFoundError: io/netty/util/internal/CleanerJava9$2
java.lang.NoClassDefFoundError: io/netty/util/internal/CleanerJava9$2
	at io.netty.util.internal.CleanerJava9.freeDirectBufferPrivileged(CleanerJava9.java:98) ~[netty-common-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.util.internal.CleanerJava9.freeDirectBuffer(CleanerJava9.java:93) ~[netty-common-4.1.79.Final.jar!/:4.1.79.Final]
	at io.netty.util.internal.PlatformDependent.freeDirectBuffer(PlatformDependent.java:523) ~[netty-common-4.1.79.Final.jar!/:4.1.79.Final]

### Reproduction Steps

When the queue is empty, there is an error since no messages are received after they are sent. If we restart the pod once again, it will consume the queue's pending messages and then throw an exception if it finds the queue is empty.

### AWS Java SDK version used

2.17.295

### JDK version used

11

### Operating System and version

Platform:"linux/amd64"

**Comment by sashjena** [comment]
Failing scenario : when pod gets restarted, pod is able to pull all available messages and process as well. and if we publish the message immediately , that also gets processed, but when there is a empty poll from pod. i.e, no available messages in the queue during the poll, some how SQS connections is getting closed in the pod and from that point on words it is not able to process any messages from SQS. So we can conclude that , there is no network or encryption/decryption issue . it is some java api issue .

**Comment by debora-ito** [comment]
`java.lang.NoClassDefFoundError` usually indicates a class is missing from your classpath. Have you fixed this error?
I'm asking this first because maybe this issue is leading to the other error.

**Comment by sashjena** [comment]
Thanks a  lot debora-ito for looking into this.

**Comment by debora-ito** [comment]
So you're not seeing the `NoClassDefFoundError ` anymore?

"Acquire operation took longer than 90000 milliseconds" indicates that a queued request timed out while waiting for a connection to become available from the connection pool. This can happen if all your connections (up to the number defined by the max connections config) are being used and no connection became available before the timeout. 

I suggest you to enable the SDK client-side metrics to have an insight on the application performance. If you are constantly getting more concurrent connections than maxConnections then consider increasing maxConnections.

To enable the client-side metrics check the SDK Dev Guide here: https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/generating-sdk-metrics.html.
This blog post explains how to interpret the metrics: https://aws.amazon.com/blogs/developer/tuning-the-aws-sdk-for-java-to-improve-resiliency/

**Comment by sashjena** [comment]
Thank a lot Debora for looking into this.
We have enabled sdk logs.

**Comment by tamandal1** [comment]
I am from @sashjena team. Those NoClassDefFoundErrors are originating from netty client library. No dependency in our code overrides netty, we verified the same by removing aws sdk dependency from code. 
We are certain that those NoClassDefFoundErrors are suppressing the actual errors, but unable to get around it.

`software.amazon.awssdk:sqs:2.17.295` pulls `software.amazon.awssdk:netty-nio-client` as transitive dependency, which in-turn pulls `netty:4.1.79` as runtime dependencies 

```
+- software.amazon.awssdk:sqs:jar:2.17.295:compile
|  +- software.amazon.awssdk:aws-query-protocol:jar:2.17.295:compile
|  +- software.amazon.awssdk:protocol-core:jar:2.17.295:compile
|  +- software.amazon.awssdk:sdk-core:jar:2.17.295:compile
|  |  \- org.reactivestreams:reactive-streams:jar:1.0.4:compile
|  +- software.amazon.awssdk:auth:jar:2.17.295:compile
|  |  +- software.amazon.awssdk:json-utils:jar:2.17.295:compile
|  |  |  \- software.amazon.awssdk:third-party-jackson-core:jar:2.17.295:compile
|  |  \- software.amazon.eventstream:eventstream:jar:1.0.1:compile
|  +- software.amazon.awssdk:http-client-spi:jar:2.17.295:compile
|  +- software.amazon.awssdk:regions:jar:2.17.295:compile
|  +- software.amazon.awssdk:annotations:jar:2.17.295:compile
|  +- software.amazon.awssdk:utils:jar:2.17.295:compile
|  +- software.amazon.awssdk:aws-core:jar:2.17.295:compile
|  +- software.amazon.awssdk:metrics-spi:jar:2.17.295:compile
|  +- software.amazon.awssdk:apache-client:jar:2.17.295:runtime
|  |  +- org.apache.httpcomponents:httpclient:jar:4.5.13:compile
|  |  \- org.apache.httpcomponents:httpcore:jar:4.4.15:compile
|  \- software.amazon.awssdk:netty-nio-client:jar:2.17.295:runtime
|     +- io.netty:netty-codec-http:jar:4.1.79.Final:runtime
|     +- io.netty:netty-codec-http2:jar:4.1.79.Final:runtime
|     +- io.netty:netty-codec:jar:4.1.79.Final:runtime
|     +- io.netty:netty-transport:jar:4.1.79.Final:runtime
|     |  \- io.netty:netty-resolver:jar:4.1.79.Final:runtime
|     +- io.netty:netty-common:jar:4.1.79.Final:runtime
|     +- io.netty:netty-buffer:jar:4.1.79.Final:runtime
|     +- io.netty:netty-handler:jar:4.1.79.Final:runtime
|     |  \- io.netty:netty-transport-native-unix-common:jar:4.1.79.Final:runtime
|     \- io.netty:netty-transport-classes-epoll:jar:4.1.79.Final:runtime
+- software.amazon.awssdk:s3:jar:2.17.295:compile
|  +- software.amazon.awssdk:aws-xml-protocol:jar:2.17.295:compile
|  +- software.amazon.awssdk:arns:jar:2.17.295:compile
|  \- software.amazon.awssdk:profiles:jar:2.17.295:compile
```

Same thing when we explicitly add `software.amazon.awssdk:netty-nio-client:jar:2.17.295` dependency!

```
+- software.amazon.awssdk:sqs:jar:2.17.295:compile
|  +- software.amazon.awssdk:aws-query-protocol:jar:2.17.295:compile
|  +- software.amazon.awssdk:protocol-core:jar:2.17.295:compile
|  +- software.amazon.awssdk:sdk-core:jar:2.17.295:compile
|  +- software.amazon.awssdk:auth:jar:2.17.295:compile
|  |  +- software.amazon.awssdk:json-utils:jar:2.17.295:compile
|  |  |  \- software.amazon.awssdk:third-party-jackson-core:jar:2.17.295:compile
|  |  \- software.amazon.eventstream:eventstream:jar:1.0.1:compile
|  +- software.amazon.awssdk:http-client-spi:jar:2.17.295:compile
|  +- software.amazon.awssdk:regions:jar:2.17.295:compile
|  +- software.amazon.awssdk:annotations:jar:2.17.295:compile
|  +- software.amazon.awssdk:utils:jar:2.17.295:compile
|  +- software.amazon.awssdk:aws-core:jar:2.17.295:compile
|  +- software.amazon.awssdk:metrics-spi:jar:2.17.295:compile
|  \- software.amazon.awssdk:apache-client:jar:2.17.295:runtime
|     +- org.apache.httpcomponents:httpclient:jar:4.5.13:compile
|     \- org.apache.httpcomponents:httpcore:jar:4.4.15:compile
+- software.amazon.awssdk:netty-nio-client:jar:2.17.295:compile
|  +- io.netty:netty-codec-http:jar:4.1.79.Final:compile
|  +- io.netty:netty-codec-http2:jar:4.1.79.Final:compile
|  +- io.netty:netty-codec:jar:4.1.79.Final:compile
|  +- io.netty:netty-transport:jar:4.1.79.Final:compile
|  |  \- io.netty:netty-resolver:jar:4.1.79.Final:compile
|  +- io.netty:netty-common:jar:4.1.79.Final:compile
|  +- io.netty:netty-buffer:jar:4.1.79.Final:compile
|  +- io.netty:netty-handler:jar:4.1.79.Final:compile
|  |  \- io.netty:netty-transport-native-unix-common:jar:4.1.79.Final:compile
|  +- io.netty:netty-transport-classes-epoll:jar:4.1.79.Final:compile
|  +- org.reactivestreams:reactive-streams:jar:1.0.4:compile
|  \- org.slf4j:slf4j-api:jar:1.7.36:compile
+- software.amazon.awssdk:s3:jar:2.17.295:compile
|  +- software.amazon.awssdk:aws-xml-protocol:jar:2.17.295:compile
|  +- software.amazon.awssdk:arns:jar:2.17.295:compile
|  \- software.amazon.awssdk:profiles:jar:2.17.295:compile
```

**Comment by debora-ito** [comment]
> We are certain that those NoClassDefFoundErrors are suppressing the actual errors, but unable to get around it.

This is my guess too.

I double-checked: Java SDK `2.17.295` uses netty `4.1.77.Final`
https://github.com/aws/aws-sdk-java-v2/blob/2.17.295/pom.xml#L105
So something in the pom is overriding the netty version, probably netty is a transitive dependency of another dependency. You can check how maven is resolving the module versions by running the verbose option -
```
 mvn dependency:tree -Dverbose=true
```

But regardless of the above, using netty `4.1.79.Final`  is not the cause of the NoClassDefFound errors. I don't have other suggestions on how to fix the classpath issue besides cleaning up the maven local repository and force the download of the dependencies again. Have you tried it?

**Comment by sashjena** [comment]
Thanks a lot Debora for your response.We attempted to forcefully add the most recent netty version, but there was no positive result.

And Few suggestion provided by AWS team, Solution / Workaround 1:
GitHub Issue: [https://github.com/netty/netty/issues/7474 ](https://github.com/netty/netty/issues/7474)
Title: sslhandler throws java.nio.channels.ClosedChannelException #7474
netty.io version: 4.1.19
> add `-Djsse.enableSNIExtension=true JAVA_OPTS`.

Solution / Workaround 2:
GtiHub Issue: [https://github.com/aws/aws-sdk-java-v2/issues/2146 ](https://github.com/aws/aws-sdk-java-v2/issues/2146) â€” same failure with the same net.io version 4.1.79
> "After further investigation, it's appeared to be an issue with LocalStack version used in our tests (0.11.1). Sometimes connection just hung. After upgrade to 0.12.2 everything works as expected."

Solution / Workaround 3:
Solution: Configure the `maxIdleTime` for the connection pool
Source: [https://github.com/reactor/reactor-netty/issues/1318#issuecomment-702668918 ](https://github.com/reactor/reactor-netty/issues/1318#issuecomment-702668918)

Solution / Workaround 4:
Two Solutions related to configuring the `maxIdleTime`
[https://github.com/reactor/reactor-netty/issues/2498#issuecomment-1263318681 ](https://github.com/reactor/reactor-netty/issues/2498#issuecomment-1263318681)
[https://github.com/reactor/reactor-netty/issues/1318#issuecomment-702668918 ](https://github.com/reactor/reactor-netty/issues/1318#issuecomment-702668918)

> If you decide to stay with the default keepAliveTimeout or any other value different than -1, then you have to configure the maxIdleTime on the Reactor Netty's connection pool to be less than the value on the server.
> If you decide to configure keepAliveTimeout with -1 then you are OK and no need to configure the connection pool.

Solution / Workaround 5:
GitHub Issue: [https://github.com/reactor/reactor-netty/issues/2499 ](https://github.com/reactor/reactor-netty/issues/2499)
Title: io.netty.handler.ssl.SslClosedEngineException: SSLEngine closed already #2499
netty.io version: 4.1.79
> investigate if the server is configured using a keep-alive timer which closes client connections on inactivity. And if it's the case, then you can try to apply the work around of using a lower idle timer on the reactor-netty http client. Source: [https://github.com/reactor/reactor-netty/issues/2499#issuecomment-1263365424 ](https://github.com/reactor/reactor-netty/issues/2499#issuecomment-1263365424)

We tried the above options suggested by AWS . Those suggest client side configuration changes around netty timeout. None gave us any positive outcome.
We played around with the following netty-io configuration, many of those suggestions given, point to these client side configurations.

NettyNioAsyncHttpClient.builder()
.connectionAcquisitionTimeout(Duration.ofSeconds(90))
.connectionMaxIdleTime(Duration.ofSeconds(30))
.connectionTimeout(Duration.ofSeconds(60))
.readTimeout(Duration.ofSeconds(60))

NOTE: Same container image works flawlessly on other environments except one of our dev env So issue is most likely not on the client side!

**Comment by debora-ito** [comment]
Hi @sashjena these are all suggestions to work around the TimeoutException, but you should look into fixing the NoClassDefFound first.

Fixing the NoClassDefFound will depend on you application environment, it's not an issue with the SDK itself so we can't give you expert suggestions. The fact that the same code works fine in other environments reinforces that.

If cleaning the local maven repository didn't work, you should search for answers on the internet for cases that match your use case with EKS pods.
